### ISSUE TITLE
Speed test fails ecdh tests

### ISSUE THREAD

**Issue by geobon** [openssl/openssl#12867]
Pulled the following version today:

MAJOR=3
MINOR=0
PATCH=0
PRE_RELEASE_TAG=alpha7-dev
BUILD_METADATA=
RELEASE_DATE=""
SHLIB_VERSION=3

It built fine.  Was going through the speed tests.  RSA tests worked but when I got to the ECDH tests I get:

```
openssl speed  -elapsed  -async_jobs 36 ecdh
You have chosen to measure elapsed time instead of user CPU time.
Unhandled error in the error queue during ECDH init.
C0A1FEF7FF7F0000:error::digital envelope routines:evp_generic_fetch:fetch failed:crypto/evp/evp_fetch.c:319:Global default library context, Algorithm (secp160r1), Properties (<null>)
Doing 160 bits  ecdh's for 10s: 65618561 160-bits ECDH ops in 10.00s
WARNING: the error queue contains previous unhandled errors.
C0A1FEF7FF7F0000:error::digital envelope routines:EVP_PKEY_derive:operation not supported for this keytype:crypto/evp/exchange.c:423:
C0A1FEF7FF7F0000:error::digital envelope routines:EVP_PKEY_derive:operation not supported for this keytype:crypto/evp/exchange.c:423:
C0A1FEF7FF7F0000:error::digital envelope routines:EVP_PKEY_derive:operation not supported for this keytype:crypto/evp/exchange.c:423:

...
Unhandled error in the error queue during ECDH init.
C0A1FEF7FF7F0000:error::digital envelope routines:evp_generic_fetch:fetch failed:crypto/evp/evp_fetch.c:319:Global default library context, Algorithm (prime192v1), Properties (<null>)
```
Nothing in the config data stands out, says ecdh is enabled.  This is just using the plain old ordinary built-in speed test.  Does the same thing in sync mode.

**Comment by geobon** [comment]
Also this is Ubuntu kernel  5.4.0-47-generic #51 if that might make any difference.

**Comment by slontis** [comment]
@levitte or @mattcaswell 

I was just debugging an issue with the eddsa speed test.. (This code was added this year)

The offending code does this

```
for (i = 0; i < max; ++i) {
    ctx[i] = EVP_MD_CTX_new();
    ctx2[i] = EVP_MD_CTX_new();

    // do a key gen to produce a key.
    …
    EVP_PKEY_keygen(genctx, &key);

    EVP_DigestSignInit(ctx[I],.. , key);
    EVP_DigestVerifyInit(ctx2[I],.. , key); // At this point the ref count of the key is 3

    EVP_PKEY_free(key); //ref count is now 2
X:
}

// Do some operation stuff in a loop
….

// Free and fall over.. with a bad reference count..
for (i = 0; i < max; ++i) {
    EVP_MD_CTX_free(ctx[i]);
    EVP_MD_CTX_free(ctx2[i]);
}
```

The issue is fixed by setting
```
key = NULL;
```
At point X.

This key is being freed but then the freed object was being passed to the next loop EVP_PKEY_keygen() call..

Does this sound like a valid fix? If you pass in an pkey to the keygen it will call pkey_set_type() -> evp_pkey_free_it() which will trash an object that is ref counted..

Shane

**Comment by t8m** [comment]
The `key = NULL;` is IMO right thing to do because we do not hold any reference after the free() call anymore.

As for the other part - whether it is right or not to trash object which might have references elsewhere with the EVP_PKEY_keygen() that's a question. Could this code work fine with 1.1.1 (of course with dropping or rather moving the EVP_PKEY_free() call outside the loop instead)?

**Comment by slontis** [comment]
The offending bit of code was only added in July 2020...

**Comment by t8m** [comment]
Yes, but the question still is - would it work on 1.1.1 or would it be broken in a similar way?

**Comment by slontis** [comment]
Will need to try it out..

**Comment by slontis** [comment]
Yep it gets confused on 111 as expected also..

On the second keygen it ends up still with a ref count of 2. (which means it will get it wrong when it tries shutting down).
This is a bug, will supply a fix - I dont know if this is the cause of the error in this issue..

**Comment by slontis** [comment]
Other calls to EVP_PKEY_free() either set it to NULL afterwards, or are in a loop where the variable is inited inside the loop.

**Comment by slontis** [comment]
it probably was not tested with -async_jobs when the code was added.

**Comment by slontis** [comment]
@geobon can you see if the fix in #12975 helps at all.

**Comment by t8m** [comment]
Fixed in alpha8.

LD_LIBRARY_PATH=. apps/openssl speed  -elapsed  -async_jobs 36 ecdh
You have chosen to measure elapsed time instead of user CPU time.
Doing 160 bits  ecdh's for 10s: 45845 160-bits ECDH ops in 10.00s
Doing 192 bits  ecdh's for 10s: 34684 192-bits ECDH ops in 10.00s
Doing 224 bits  ecdh's for 10s: 96079 224-bits ECDH ops in 10.00s
Doing 256 bits  ecdh's for 10s: 142746 256-bits ECDH ops in 10.00s
Doing 384 bits  ecdh's for 10s: 8659 384-bits ECDH ops in 10.00s
Doing 521 bits  ecdh's for 10s: 20325 521-bits ECDH ops in 10.00s
Doing 256 bits  ecdh's for 10s: 18550 256-bits ECDH ops in 10.00s
Doing 256 bits  ecdh's for 10s: 18744 256-bits ECDH ops in 10.00s
Doing 384 bits  ecdh's for 10s: 8403 384-bits ECDH ops in 10.00s
Doing 384 bits  ecdh's for 10s: 8288 384-bits ECDH ops in 10.00s
Doing 512 bits  ecdh's for 10s: 5642 512-bits ECDH ops in 10.00s
Doing 512 bits  ecdh's for 10s: 5621 512-bits ECDH ops in 10.00s
Doing 253 bits  ecdh's for 10s: 217588 253-bits ECDH ops in 10.00s
Doing 448 bits  ecdh's for 10s: 14572 448-bits ECDH ops in 10.01s
version: 3.0.0-alpha8
built on: built on: Thu Nov  5 00:00:00 2020 UTC
options:bn(64,64) md2(char) rc4(16x,int) des(int) aes(partial) blowfish(ptr) 
compiler: gcc -fPIC -pthread -m64 -Wa,--noexecstack -Wall -O3 -O2 -g -pipe -Wall -Werror=format-security -Wp,-D_FORTIFY_SOURCE=2 -Wp,-D_GLIBCXX_ASSERTIONS -fexceptions -fstack-protector-strong -grecord-gcc-switches -specs=/usr/lib/rpm/redhat/redhat-hardened-cc1 -specs=/usr/lib/rpm/redhat/redhat-annobin-cc1 -m64 -mtune=generic -fasynchronous-unwind-tables -fstack-clash-protection -fcf-protection -Wa,--noexecstack -Wa,--generate-missing-build-notes=yes -specs=/usr/lib/rpm/redhat/redhat-hardened-ld -DOPENSSL_USE_NODELETE -DL_ENDIAN -DOPENSSL_PIC -DOPENSSL_BUILDING_OPENSSL -DZLIB -DNDEBUG -DPURIFY -DDEVRANDOM="\"/dev/urandom\"" -DSYSTEM_CIPHERS_FILE="/etc/crypto-policies/back-ends/openssl.config"
CPUINFO: OPENSSL_ia32cap=0x7ffaf3ffffebffff:0x29c67af
                              op      op/s
 160 bits ecdh (secp160r1)   0.0002s   4584.5
 192 bits ecdh (nistp192)   0.0003s   3468.4
 224 bits ecdh (nistp224)   0.0001s   9607.9
 256 bits ecdh (nistp256)   0.0001s  14274.6
 384 bits ecdh (nistp384)   0.0012s    865.9
 521 bits ecdh (nistp521)   0.0005s   2032.5
 256 bits ecdh (brainpoolP256r1)   0.0005s   1855.0
 256 bits ecdh (brainpoolP256t1)   0.0005s   1874.4
 384 bits ecdh (brainpoolP384r1)   0.0012s    840.3
 384 bits ecdh (brainpoolP384t1)   0.0012s    828.8
 512 bits ecdh (brainpoolP512r1)   0.0018s    564.2
 512 bits ecdh (brainpoolP512t1)   0.0018s    562.1
 253 bits ecdh (X25519)   0.0000s  21758.8
 448 bits ecdh (X448)   0.0007s   1455.7
